File: android/src/main/java/com/svgaplayer/RNSvgaPlayerManager.java
package com.svgaplayer;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.infer.annotation.Assertions;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewManagerDelegate;
import com.facebook.react.uimanager.annotations.ReactProp;
import com.facebook.react.viewmanagers.SvgaPlayerViewManagerDelegate;
import com.facebook.react.viewmanagers.SvgaPlayerViewManagerInterface;
import com.opensource.svgaplayer.SVGAParser;
import com.opensource.svgaplayer.SVGAVideoEntity;
import com.svgaplayer.events.TopFinishedEvent;
import com.svgaplayer.events.TopFrameEvent;
import com.svgaplayer.events.TopPercentage;

import java.net.URL;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

@ReactModule(name = RNSvgaPlayerManager.REACT_CLASS)
public class RNSvgaPlayerManager extends SimpleViewManager<RCTSVGAImageView> implements SvgaPlayerViewManagerInterface<RCTSVGAImageView> {
    public static final String REACT_CLASS = "SvgaPlayerView";
    private final ViewManagerDelegate<RCTSVGAImageView> mDelegate;

    @NonNull
    @Override
    public String getName() {
        return REACT_CLASS;
    }

    public RNSvgaPlayerManager() {
        mDelegate = new SvgaPlayerViewManagerDelegate<>(this);
    }

    @Nullable
    @Override
    protected ViewManagerDelegate<RCTSVGAImageView> getDelegate() {
        return mDelegate;
    }

    @NonNull
    @Override
    public RCTSVGAImageView createViewInstance(@NonNull ThemedReactContext c) {
        return new RCTSVGAImageView(c, null, 0);
    }
    @ReactProp(name = "source")
    public void setSource(final RCTSVGAImageView view, String source) {
        Context context = view.getContext();
        assert source != null;
        if (source.startsWith("http") || source.startsWith("https")) {
            try {
                new SVGAParser(context).parse(new URL(source), new SVGAParser.ParseCompletion() {
                    @Override
                    public void onError() {

                    }

                    @Override
                    public void onComplete(SVGAVideoEntity videoItem) {
                        view.setVideoItem(videoItem);
                        view.startAnimation();
                    }

                });
            } catch (Exception e) {
            }
        } else {
            try {
                new SVGAParser(context).decodeFromAssets(source, new SVGAParser.ParseCompletion() {
                    @Override
                    public void onError() {

                    }

                    @Override
                    public void onComplete(SVGAVideoEntity videoItem) {
                        view.setVideoItem(videoItem);
                        view.startAnimation();
                    }
                });
            } catch (Exception e) {
            }
        }
    }

    @ReactProp(name = "loops", defaultInt = 0)
    public void setLoops(RCTSVGAImageView view, int loops) {
        view.setLoops(loops);
    }

    @ReactProp(name = "clearsAfterStop", defaultBoolean = true)
    public void setClearsAfterStop(RCTSVGAImageView view, boolean clearsAfterStop) {
        view.setClearsAfterStop(clearsAfterStop);
    }

    @Override
    public void load(RCTSVGAImageView view, String source) {
        this.setSource(view,source);
    }

    @ReactProp(name = "currentState")
    public void setCurrentState(RCTSVGAImageView view, String currentState) {
        view.currentState = currentState;
        switch (currentState) {
            case "start":
                view.startAnimation();
                break;
            case "pause":
                view.pauseAnimation();
                break;
            case "stop":
                view.stopAnimation();
                break;
            case "clear":
                view.stopAnimation(true);
                break;
            default:
                break;
        }
    }

    @ReactProp(name = "toFrame", defaultFloat = -1.0f)
    public void setToFrame(RCTSVGAImageView view, float toFrame) {
        if (toFrame < 0) {
            return;
        }
        view.stepToFrame((int) toFrame, view.currentState.equals("play"));
    }

    @ReactProp(name = "toPercentage", defaultFloat = -1.0f)
    public void setToPercentage(RCTSVGAImageView view, float toPercentage) {
        if (toPercentage < 0) {
            return;
        }
        view.stepToPercentage((double) toPercentage, view.currentState.equals("play"));
    }

    @Override
    public void receiveCommand(@NonNull RCTSVGAImageView root, String commandId, @Nullable ReadableArray args) {
        super.receiveCommand(root, commandId, args);
        if ("load".equals(commandId)) {
            Assertions.assertNotNull(args);
            assert args != null;
            load(root, args.getString(0));
        } else  if ("startAnimation".equals(commandId)) {
             startAnimation(root);
        } else  if ("pauseAnimation".equals(commandId)) {
           pauseAnimation(root);

        }else   if ("stopAnimation".equals(commandId)) {
            stopAnimation(root);
        }
    }

    @Override
    public void startAnimation(RCTSVGAImageView view) {
        view.startAnimation();
    }

    @Override
    public void pauseAnimation(RCTSVGAImageView view) {
        view.pauseAnimation();
    }

    @Override
    public void stopAnimation(RCTSVGAImageView view) {
        view.stopAnimation(true);
    }
//    @Override
//    public void stepToFrame(RCTSVGAImageView view, float toFrame, boolean andPlay) {
//        if (toFrame < 0) {
//            return;
//        }
//        view.stepToFrame((int) toFrame, view.currentState.equals("play"));
//
//    }
//    @Override
//    public void stepToPercentage(RCTSVGAImageView view, float toPercentage, boolean andPlay) {
//        if (toPercentage < 0) {
//            return;
//        }
//        view.stepToPercentage((double) toPercentage, view.currentState.equals("play"));
//
//    }
@Override
public Map<String, Object> getExportedCustomDirectEventTypeConstants() {
    Map<String, Object> export = super.getExportedCustomDirectEventTypeConstants();
    if (export == null) {
        export = MapBuilder.newHashMap();
    }
    // Default events but adding them here explicitly for clarity
    export.put(TopFrameEvent.EVENT_NAME, MapBuilder.of("registrationName", "onFrame"));
    export.put(TopPercentage.EVENT_NAME, MapBuilder.of("registrationName", "onPercentage"));
    export.put(TopFinishedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onFinished"));
       return export;
}
//    @Nullable
//    @Override
//    public Map<String, Object> getExportedCustomBubblingEventTypeConstants() {
//        Map<String, Object> map = new HashMap<>();
//        Map<String, Object> frameBubblingMap = new HashMap<>();
//        frameBubblingMap.put("phasedRegistrationNames", new HashMap<String, String>() {{
//            put("bubbled", "onFrame");
////            put("captured", "onFrameCapture");
//        }});
//
//        map.put("onFrame", frameBubblingMap);
//        Map<String, Object> percentageBubblingMap = new HashMap<>();
//        percentageBubblingMap.put("phasedRegistrationNames", new HashMap<String, String>() {{
//            put("bubbled", "onPercentage");
////            put("captured", "onFrameCapture");
//        }});
//
//        map.put("onPercentage", percentageBubblingMap);
//        Map<String, Object> finishBubblingMap = new HashMap<>();
//        finishBubblingMap.put("phasedRegistrationNames", new HashMap<String, String>() {{
//            put("bubbled", "onFinished");
////            put("captured", "onFrameCapture");
//        }});
//        map.put("onFinished",finishBubblingMap);
//        return map;
//    }

}

File: android/src/main/java/com/svgaplayer/RNSvgaPlayerManager.kt
package com.svgaplayer

import android.util.Log
import android.widget.ImageView
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.common.MapBuilder
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.UIManagerHelper
import com.facebook.react.uimanager.ViewManagerDelegate
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.viewmanagers.RNSvgaPlayerManagerDelegate
import com.facebook.react.viewmanagers.RNSvgaPlayerManagerInterface
import com.opensource.svgaplayer.SVGACache
import com.opensource.svgaplayer.SVGAParser
import com.opensource.svgaplayer.SVGAVideoEntity
import com.opensource.svgaplayer.utils.SVGARange
import com.svgaplayer.events.TopErrorEvent
import com.svgaplayer.events.TopLoadedEvent
import java.io.File
import java.io.FileInputStream
import java.net.URL


@ReactModule(name = RNSvgaPlayerManager.REACT_CLASS)
class RNSvgaPlayerManager() : SimpleViewManager<RNSvgaPlayer>(), RNSvgaPlayerManagerInterface<RNSvgaPlayer> {

  companion object {
    const val REACT_CLASS = "RNSvgaPlayer"
  }
  private val mDelegate: ViewManagerDelegate<RNSvgaPlayer> = RNSvgaPlayerManagerDelegate(this)

  override fun getName(): String = REACT_CLASS

  override fun getDelegate(): ViewManagerDelegate<RNSvgaPlayer>? = mDelegate

  override fun createViewInstance(c: ThemedReactContext): RNSvgaPlayer {
    return RNSvgaPlayer(c, null, 0)
  }
  @ReactProp(name = "source")
  override fun setSource(view: RNSvgaPlayer, source: String?) {
    val context = view.context
    source?.let {
      val parseCompletion = object : SVGAParser.ParseCompletion {
        override fun onError() {
          view.setVideoItem(null)
          view.clear()

          val errorData = Arguments.createMap()
          errorData.putString("error", "Failed to load SVGA : $it")
          val surfaceId = UIManagerHelper.getSurfaceId(context)
          val errorEvent = TopErrorEvent(surfaceId, view.id, errorData)
          val dispatcher = UIManagerHelper.getEventDispatcherForReactTag(context as ThemedReactContext, view.id)
          dispatcher?.dispatchEvent(errorEvent)
        }

        override fun onComplete(videoItem: SVGAVideoEntity) {
          view.setVideoItem(videoItem)

          // 触发 onLoaded 事件
          val loadedData = Arguments.createMap()
          val surfaceId = UIManagerHelper.getSurfaceId(context)
          val loadedEvent = TopLoadedEvent(surfaceId, view.id, loadedData)
          val dispatcher = UIManagerHelper.getEventDispatcherForReactTag(context as ThemedReactContext, view.id)
           dispatcher?.dispatchEvent(loadedEvent)

          if(view.autoPlay){
            view.startAnimationSafely()
          }
        }
      }

      when {
        it.startsWith("http") || it.startsWith("https") -> {
          SVGAParser(context).decodeFromURL(URL(it), parseCompletion)
        }
        it.startsWith("file://") -> {
          // 移除 file:// 前缀，获取实际文件路径
          val filePath = it.removePrefix("file://")
          val file = File(filePath)

          if (file.exists() && file.isFile) {
            val inputStream = FileInputStream(file)
            val cacheKey = SVGACache.buildCacheKey(it)
            SVGAParser(context).decodeFromInputStream(inputStream, cacheKey, parseCompletion)
          } else {
            parseCompletion.onError()
          }
        }
        else -> {
          Log.d("RNSvgaPlayerManager", "Loading from assets: $it")
          SVGAParser(context).decodeFromAssets(it, parseCompletion)
        }
      }
    }
  }
  @ReactProp(name = "loops", defaultInt = 0)
  override fun setLoops(view: RNSvgaPlayer, loops: Int) {
    view.loops = loops
  }
  @ReactProp(name = "clearsAfterStop", defaultBoolean = true)
  override fun setClearsAfterStop(view: RNSvgaPlayer, clearsAfterStop: Boolean) {
    view.clearsAfterDetached = clearsAfterStop
    view.clearsAfterStop = clearsAfterStop
  }
  @ReactProp(name = "autoPlay", defaultBoolean = true)
  override fun setAutoPlay(view: RNSvgaPlayer, autoPlay: Boolean) {
    view.autoPlay = autoPlay
  }
  @ReactProp(name = "align")
  override fun setAlign(view: RNSvgaPlayer, align: String?) {
    val scaleType = when (align) {
      "bottom" -> ImageView.ScaleType.FIT_END
      "top" -> ImageView.ScaleType.FIT_START
      "center" -> ImageView.ScaleType.FIT_CENTER
      else -> ImageView.ScaleType.FIT_CENTER
    }

    view.scaleType = scaleType
  }


  override fun receiveCommand(root: RNSvgaPlayer, commandId: String, args: ReadableArray?) {
    super.receiveCommand(root, commandId, args)
    when (commandId) {
      "startAnimation" -> startAnimation(root)
      "stopAnimation" -> stopAnimation(root)
      "pauseAnimation" -> pauseAnimation(root)
      "stepToFrame" -> {
        val frame = args?.getInt(0) ?: 0
        val andPlay = args?.getBoolean(1) ?: false
        stepToFrame(root, frame, andPlay)
      }
      "stepToPercentage" -> {
        val percentage = args?.getInt(0) ?: 0
        val andPlay = args?.getBoolean(1) ?: false
        stepToPercentage(root, percentage, andPlay)
      }
      "startAnimationWithRange" -> {
        val location = args?.getInt(0) ?: 0
        val length = args?.getInt(1) ?: 0
        val reverse = args?.getBoolean(2) ?: false
        startAnimationWithRange(root, location, length, reverse)
      }
      else -> {
        Log.w(REACT_CLASS, "Unknown command: $commandId")
      }
    }
  }

  override fun startAnimation(view: RNSvgaPlayer) {
    view.startAnimationSafely()
  }

  override fun stopAnimation(view: RNSvgaPlayer) {
    view.stopAnimation(view.clearsAfterStop)
  }

//  override fun getExportedCustomBubblingEventTypeConstants(): Map<String, Any>? {
//    val export = super.getExportedCustomDirectEventTypeConstants()?.toMutableMap()
//      ?: mutableMapOf<String, Any>()
//
//    export[TopErrorEvent.EVENT_NAME] = mapOf("registrationName" to "onError")
//    export[TopFinishedEvent.EVENT_NAME] = mapOf("registrationName" to "onFinished")
//    export[TopLoadedEvent.EVENT_NAME] = mapOf("registrationName" to "onLoaded")
//    export[TopFrameEvent.EVENT_NAME] = mapOf("registrationName" to "onFrameChanged")
//    export[TopPercentageEvent.EVENT_NAME] = mapOf("registrationName" to "onPercentageChanged")
//
//    return export
//  }
  override fun getExportedCustomBubblingEventTypeConstants(): MutableMap<String, Any>? {
  val map = MapBuilder.builder<String, Any>()
    .put("topError", MapBuilder.of("registrationName", "onError"))
    .put("topFinished", MapBuilder.of("registrationName", "onFinished"))
    .put("topLoaded", MapBuilder.of("registrationName", "onLoaded"))
    .put("topFrameChanged", MapBuilder.of("registrationName", "onFrameChanged"))
    .put("topPercentageChanged", MapBuilder.of("registrationName", "onPercentageChanged"))
    .build()
  return map
  }

  override fun pauseAnimation(view: RNSvgaPlayer?) {
    view?.pauseAnimation()
  }

  override fun stepToFrame(view: RNSvgaPlayer?, frame: Int, andPlay: Boolean) {
    view?.stepToFrame(frame, andPlay);
  }

  override fun stepToPercentage(view: RNSvgaPlayer?, percentage: Int, andPlay: Boolean) {
    view?.stepToFrame(percentage, andPlay);
  }

  override fun startAnimationWithRange(
    view: RNSvgaPlayer?,
    location: Int,
    length: Int,
    reverse: Boolean
  ) {
    var  range = SVGARange(location, length)
    view?.startAnimation(range,reverse)
  }
}
